plugins/inputs/snmp2/snmp.go

```diff
@@ -0,0 +1,624 @@
+	return description
+}
+
+// Gather retrieves all the configured fields and tables.
+// Any error encountered does not halt the process. The errors are accumulated
+// and returned at the end.
+func (s *Snmp2) Gather(acc telegraf.Accumulator) error {
+	var errs Errors
+	for _, agent := range s.Agents {
+		gs, err := s.getConnection(agent)
+		if err != nil {
+			errs = append(errs, Errorf(err, "agent %s", agent))
+			continue
+		}
+
+		// First is the top-level fields. We treat the fields as table prefixes with an empty index.
+		t := Table{
+			Name:   s.Name,
+			Fields: s.Fields,
+		}
+		topTags := map[string]string{}
+		if err := s.gatherTable(acc, gs, t, topTags, false); err != nil {
+			errs = append(errs, Errorf(err, "agent %s", agent))
+		}
+
+		// Now is the real tables.
+		for _, t := range s.Tables {
+			if err := s.gatherTable(acc, gs, t, topTags, true); err != nil {
+				errs = append(errs, Errorf(err, "agent %s", agent))
+			}
+		}
+	}
+
+	if errs == nil {
+		return nil
+	}
+	return errs
+}
+
+func (s *Snmp2) gatherTable(acc telegraf.Accumulator, gs snmpConnection, t Table, topTags map[string]string, walk bool) error {
+	rt, err := t.Build(gs, walk)
+	if err != nil {
+		return err
+	}
+
+	for _, tr := range rt.Rows {
+		if !walk {
+			// top-level table. Add tags to topTags.
+			for k, v := range tr.Tags {
+				topTags[k] = v
+			}
+		} else {
+			// real table. Inherit any specified tags.
+			for _, k := range t.InheritTags {
+				if v, ok := topTags[k]; ok {
+					tr.Tags[k] = v
+				}
+			}
+		}
+		if _, ok := tr.Tags["agent_host"]; !ok {
+			tr.Tags["agent_host"] = gs.Host()
+		}
+		acc.AddFields(rt.Name, tr.Fields, tr.Tags, rt.Time)
+	}
+
+	return nil
+}
+
+// Build retrieves all the fields specified in the table and constructs the RTable.
+func (t Table) Build(gs snmpConnection, walk bool) (*RTable, error) {
+	rows := map[string]RTableRow{}
+
+	tagCount := 0
+	for _, f := range t.Fields {
+		if f.IsTag {
+			tagCount++
+		}
+
+		if len(f.Oid) == 0 {
+			return nil, fmt.Errorf("cannot have empty OID")
+		}
+		var oid string
+		if f.Oid[0] == '.' {
+			oid = f.Oid
+		} else {
+			// make sure OID has "." because the BulkWalkAll results do, and the prefix needs to match
+			oid = "." + f.Oid
+		}
+
+		// ifv contains a mapping of table OID index to field value
+		ifv := map[string]interface{}{}
+
+		if !walk {
+			// This is used when fetching non-table fields. Fields configured a the top
+			// scope of the plugin.
+			// We fetch the fields directly, and add them to ifv as if the index were an
+			// empty string. This results in all the non-table fields sharing the same
+			// index, and being added on the same row.
+			if pkt, err := gs.Get([]string{oid}); err != nil {
+				return nil, Errorf(err, "performing get")
+			} else if pkt != nil && len(pkt.Variables) > 0 && pkt.Variables[0].Type != gosnmp.NoSuchObject {
+				ent := pkt.Variables[0]
+				ifv[ent.Name[len(oid):]] = fieldConvert(f.Conversion, ent.Value)
```

Comment:
The `ifv` variable is a map of the values in each row for a single column. The map's key is the field index. The field index is the portion of the OID after the table prefix.

Lets say you have the following data:
```
.1.1.1.2.0: "aaa"
.1.1.1.2.1: "bbb"
.1.1.1.3.0: "xxx"
.1.1.1.3.1: "yyy"
```
Your fields are `foo=.1.1.1.2` and `bar=.1.1.1.3`.
Our end result that we want to achieve is:
```
foo="aaa",bar="xxx"
foo="bbb",bar="yyy"
````

So in the above, your table is `.1.1.1`.
One column is `.1.1.1.2` with row values `.1.1.1.2.0=aaa` and `.1.1.1.2.1=bbb`. The field indexes are `.0` and `.1`, and `ifv` becomes `map{".0":"aaa", ".1":"bbb"}`.
The next column is `.1.1.1.3`, with row values `.1.1.1.3.0=xxx` and `.1.1.1.3.1=yyy`, and `ifv` becomes `map{".0":"xxx", ".1":"yyy"}`.
The result is that you have a bunch of `ifv` variables, one for each column, but they all share the same keys (which allows you to reorganize the data into rows instead of columns).

```
map{".0":"aaa", ".1":"bbb"} // prefix=".1.1.1.2"
map{".0":"xxx", ".1":"yyy"} // prefix=".1.1.1.3"
```
But if you just did `ifv[ent.Name]`, you'd end up with
```
map{".1.1.1.2.0":"aaa", ".1.1.1.2.1":"bbb"}
map{".1.1.1.3.0":"xxx", ".1.1.1.3.1":"yyy"}
```
Which is a lot harder to turn into our end result:
```
map{".1.1.1.2":"aaa", ".1.1.1.3":"xxx"} // index=".0"
map{".1.1.1.2":"bbb", ".1.1.1.3":"yyy"} // index=".1"
```

And no, you can't just use the last digit in the OID as the index. What if you had the following data instead:
```
.1.1.1.2.4.5.6: "aaa"
.1.1.1.2.7: "bbb"
.1.1.1.3.4.5.6: "xxx"
.1.1.1.3.7: "yyy"
```
*(This is a common method of embedding strings in OIDs, known as a "string index". For example `.104.101.108.108.111` is the string `hello`)*

Your table is still `.1.1.1`, and your fields are still `foo=.1.1.1.2` and `bar=.1.1.1.3`. If you did `ifv[ent.Name]`, you'd end up with:
```
map{".1.1.1.2.4.5.6":"aaa", ".1.1.1.2.7":"bbb"}
map{".1.1.1.3.4.5.6":"xxx", ".1.1.1.3.7":"yyy"}
```
How would you then convert that into:
```
map{".1.1.1.2":"aaa", ".1.1.1.3":"xxx"} // index=".4.5.6"
map{".1.1.1.2":"bbb", ".1.1.1.3":"yyy"} // index=".7"
```

---

Now all this is really only for table data. On this specific line of the code, we aren't handling table data, we're handling the fields that were configured at the top level of the plugin. But lets say you have the configuration:
```
[[inputs.snmp2]]
    [[inputs.snmp2.field]]
        name = "f1"
        oid = ".2.2.2.0"
    [[inputs.snmp2.field]]
        name = "f2"
        oid = ".2.2.3.4.5"
    [[inputs.snmp2.field]]
        name = "f3"
        oid = ".2.2.6"
```
We don't want to output this to influxdb as
```
snmp2 f1="foo"
snmp2 f2="bar"
snmp2 f3="baz"
```
We want to output it as
```
snmp2 f1="foo",f2="bar",f3="baz"
```

We accomplish this by pretending that the index on all three of those fields is the same (specifically the empty string `""`, but anything will do as long as it's the same), and so we end up with 3 maps:
```
map{"":"foo"} // prefix="2.2.2.0"
map{"":"bar"} // prefix="2.2.2.3.4.5"
map{"":"baz"} // prefix="2.2.6"
```
And when we flip this so that we have maps of row values instead of maps of column values, we end up with:
```
map{"2.2.2.0":"foo", "2.2.2.3.4.5":"bar", "2.2.6":"baz"}
```
Doing it this way is not only to output a single row to influxdb, but it also lets us share code between non-table and table data.